CPP/DesktopToastsCppApp/DesktopNotificationManagerCompat.cpp

```diff
@@ -0,0 +1,288 @@
+    wchar_t exePath[MAX_PATH];
+    DWORD charWritten = ::GetModuleFileName(nullptr, exePath, ARRAYSIZE(exePath));
+    auto hr = charWritten > 0 ? S_OK : HRESULT_FROM_WIN32(::GetLastError());
+    if (SUCCEEDED(hr))
+    {
+        // Register the COM server
+        hr = RegisterComServer(clsid, exePath);
+        if (SUCCEEDED(hr))
+        {
+            s_registeredAumidAndComServer = true;
+        }
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterActivator()
+{
+    // Module<OutOfProc> needs a callback registered before it can be used.
+    // Since we don't care about when it shuts down, we'll pass an empty lambda here.
+    Module<OutOfProc>::Create([] {});
+
+    // If a local server process only hosts the COM object then COM expects
+    // the COM server host to shutdown when the references drop to zero.
+    // Since the user might still be using the program after activating the notification,
+    // we don't want to shutdown immediately.  Incrementing the object count tells COM that
+    // we aren't done yet.
+    Module<OutOfProc>::GetModule().IncrementObjectCount();
+
+    auto hr = Module<OutOfProc>::GetModule().RegisterObjects();
+    if (SUCCEEDED(hr))
+    {
+        s_registeredActivator = true;
+    }
+
+    return hr;
+}
+
+HRESULT DesktopNotificationManagerCompat::RegisterComServer(GUID clsid, wchar_t exePath[])
+{
+    // Turn the GUID into a string
+    OLECHAR* clsidOlechar;
+    StringFromCLSID(clsid, &clsidOlechar);
+    std::wstring clsidStr(clsidOlechar);
+    ::CoTaskMemFree(clsidOlechar);
+
+    // Create the subkey
+    // Something like SOFTWARE\Classes\CLSID\{23A5B06E-20BB-4E7E-A0AC-6982ED6A6041}\LocalServer32
+    std::wstring subKeyStr = LR"(SOFTWARE\Classes\CLSID\)" + clsidStr + LR"(\LocalServer32)";
+    LPCWSTR subKey = subKeyStr.c_str();
+
+    // Include -ToastActivated launch args on the exe
+    std::wstring exePathStr(exePath);
+    exePathStr = L"\"" + exePathStr + L"\" -ToastActivated";
+    exePath = new wchar_t[exePathStr.length() + 1];
+    wcscpy(exePath, exePathStr.c_str());
+
+    // We don't need to worry about overflow here as ::GetModuleFileName won't
+    // return anything bigger than the max file system path (much fewer than max of DWORD).
+    DWORD dataSize = static_cast<DWORD>((::wcslen(exePath) + 1) * sizeof(WCHAR));
+
+    // Register the EXE for the COM server
+    return HRESULT_FROM_WIN32(::RegSetKeyValue(
+        HKEY_CURRENT_USER,
+        subKey,
+        nullptr,
+        REG_SZ,
+        reinterpret_cast<const BYTE*>(exePath),
+        dataSize));
+}
+
+HRESULT DesktopNotificationManagerCompat::CreateToastNotifier(IToastNotifier **notifier)
+{
+    EnsureRegistered();
+
+    ComPtr<IToastNotificationManagerStatics> toastStatics;
+    HRESULT hr = Windows::Foundation::GetActivationFactory(
+        HStringReference(RuntimeClass_Windows_UI_Notifications_ToastNotificationManager).Get(),
+        &toastStatics);
+    if (SUCCEEDED(hr))
```

Comment:
nit: I really prefer early return rather than `if (SUCCEEDED(hr))` staircase. It doesn't matter much in this function but as you add more checks this becomes more and more ugly.